### DISCUSSION TITLE
Update-BTNotification with -ProgressBar option

### DISCUSSION THREAD

**Discussion by PhilZ-cwm6** [Windos/BurntToast#103] [General] [answered]
Hi,
First many thanks for this great tool !

Is there a way to update the ProgressBar without refreshing the whole toast ?
Something like `Update-BTNotification -ProgressBar $newprogress` which currently throws an error 

Best regards

**Comment by Windos** [comment]
Hey @PhilZ-cwm6, I think I've got some documentation on how to do this with `Update-BTNotification` (which you're on the right track, it is how you're meant to do it), I'll need to do some hunting to find it.

What's the error and how're you trying to use it current?

**Comment by PhilZ-cwm6** [comment]
Here's the sample test.ps1 I am trying
```
$FirstDataBinding = @{
    FirstLine = "Example Toast Heading"
    SecondLine = "This toast is still the original"
    ThirdLine = " "
}

$SecondDataBinding = @{
    SecondLine = "This toast has been updated!"
    ThirdLine = "hello"
}

Import-Module BurntToast
$Progress = New-BTProgressBar -Status "Backup progress..." -Value 0.4 -ValueDisplay "4/10"

New-BurntToastNotification -Text "FirstLine", "SecondLine", "ThirdLine" -UniqueIdentifier "ExampleToast" -DataBinding $FirstDataBinding -ProgressBar $Progress
Start-Sleep 10
$Progress = New-BTProgressBar -Status "Backup progress..." -Value 0.8 -ValueDisplay "8/10"
New-BurntToastNotification -Text "FirstLine", "SecondLine", "ThirdLine" -UniqueIdentifier "ExampleToast" -DataBinding $FirstDataBinding -ProgressBar $Progress

```

That code works perfectly, except all the toast is refreshed/recreated (as expected) when calling `New-BurntToastNotification`
When I call: `Update-BTNotification -UniqueIdentifier 'ExampleToast' -DataBinding $SecondDataBinding -ProgressBar $Progress` it would fail with:
```
Update-BTNotification : A parameter cannot be found that matches parameter name 'ProgressBar'.
At ./test.ps1:88 char:88
+ ... r 'ExampleToast' -DataBinding $SecondDataBinding -ProgressBar $Progre ...
+                                                      ~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Update-BTNotification], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Update-BTNotification
```

This is a sample test from a long backup script. Currently, I refresh the progress during backup job using the New-BurntToastNotification in a `foreach ` loop

Well, it is just perfect as it is now for me, however, when I saw the `Update-BTNotification` I was curious to push it a bit more :-)

**Reply by Windos** [reply]
Ahh! You're very close!

So those data binding for the text, also apply to the progress bar itself.

For your first toast, try this:

```powershell
$FirstDataBinding = @{
    FirstLine = "Example Toast Heading"
    SecondLine = "This toast is still the original"
    ThirdLine = " "
    ProgressStatus = 'Backup progress...'
    ProgressValue = 0.1
    ProgressDisplay = "1/10"
}

$Progress = New-BTProgressBar -Status "ProgressStatus" -Value 'ProgressValue' -ValueDisplay 'ProgressDisplay'

New-BurntToastNotification -Text "FirstLine", "SecondLine", "ThirdLine" -UniqueIdentifier "ExampleToast" -DataBinding $FirstDataBinding -ProgressBar $Progress
```

Note that the three data elements in `New-BTProgressBar` reference values in `$FirstDataBinding`.

To update it, send through an updated data binding:

```powershell
$SecondDataBinding = @{
    SecondLine = "This toast has been updated!"
    ThirdLine = "hello"
    ProgressStatus = 'Backup progress...'
    ProgressValue = 0.3
    ProgressDisplay = "4/10"
}

Update-BTNotification -UniqueIdentifier "ExampleToast" -DataBinding $SecondDataBinding
```

And continue until it's finished:

```powershell
$ThirdDataBinding = @{
    SecondLine = "This toast has been updated!"
    ThirdLine = "hello"
    ProgressStatus = 'Backup progress...'
    ProgressValue = 0.7
    ProgressDisplay = "7/10"
}

Update-BTNotification -UniqueIdentifier "ExampleToast" -DataBinding $ThirdDataBinding

$FourthDataBinding = @{
    SecondLine = "This toast has been updated!"
    ThirdLine = "hello"
    ProgressStatus = 'Backup complete'
    ProgressValue = 1
    ProgressDisplay = "10/10"
}

Update-BTNotification -UniqueIdentifier "ExampleToast" -DataBinding $FourthDataBinding
```

Random aside here in case it's handy for your code/usecase: remember you can re-use the databinding hashtable and just change the elements needed:

```powershell
DataBinding = @{
    FirstLine = "Example Toast Heading"
    SecondLine = "This toast is still the original"
    ThirdLine = " "
    ProgressStatus = 'Backup progress...'
    ProgressValue = 0.1
    ProgressDisplay = "1/10"
}

# Do the first toast

$DataBinding['ProgressValue '] = 0.2
$DataBinding['ProgressDisplay'] = "2/10"

# Do the second toast, etc.
```

**Reply by PhilZ-cwm6** [reply]
Ah, now that's great yes, reusing the Databinding is what I need. However, I am not really sure how to implement it properly in my case.

I use `New-BurntToastNotification @toastParams_some_status` through the code with all the @toastParams defined at begin of code like this:
```
# NAS BurntToastNotification params
$toastParams_NAS_Offline = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($messageNAS_Offline), ($messageNAS_Retry10m), ($timestamp).ToString()
    AppLogo = $Logo_Warn
}
$toastParams_NAS_Failed = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($messageNAS_Failed), ($timestamp).ToString()
    AppLogo = $Logo_Error
}
$toastParams_NAS_BackOnline = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    Text = ($messageNAS_Online), ($timestamp).ToString()
    AppLogo = $Logo_Success
}

# Backup BurntToastNotification params
$toastParams_Backup_Progress = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($msgBackup_Started), ($timestamp).ToString()
    AppLogo = $Logo_Progress
}
$toastParams_Backup_Started = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    Text = ($msgBackup_Started), ($timestamp).ToString()
    AppLogo = $Logo_Success
}
$toastParams_Backup_Success = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    Text = ($msgBackup_Success), ($timestamp).ToString()
    AppLogo = $Logo_Success
    Button = New-BTButton -Content 'Log files...' -Arguments $FFS_Logs
}
$toastParams_Backup_Warn = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($msgBackup_Warn), ($timestamp).ToString()
    AppLogo = $Logo_Warn
    Button = New-BTButton -Content 'Log files...' -Arguments $FFS_Logs
}
$toastParams_Backup_Error = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($msgBackup_Error), ($timestamp).ToString()
    AppLogo = $Logo_Error
    Button = New-BTButton -Content 'Log files...' -Arguments $FFS_Logs
}
$toastParams_Backup_Cancel = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($msgBackup_Cancel), ($timestamp).ToString()
    AppLogo = $Logo_Cancel
    Button = New-BTButton -Content 'Log files...' -Arguments $FFS_Logs
}
$toastParams_Backup_Unknown = @{
    UniqueIdentifier = $TargetNoticationfName
    Header = (New-BTHeader -Id 1 -Title $TargetNoticationfName)
    #Text = ($msgBackup_Unknown), ($timestamp).ToString()
    AppLogo = $Logo_Unknown
    Button = New-BTButton -Content 'Log files...' -Arguments $FFS_Logs
}
# and so on...
```

That way, I can define the text messages at start of my script for easier maintenance
The toast I want to dynamically update with progress is the `$toastParams_Backup_Progress`

Currently, the main code is started like this in a foreach loop:

```
foreach ($ffs_batch_script in $ffs_batch_scripts_array) {
# backup code and call to functions checking for drive mounts and network shares availability 

# notification of progress
    $Progress_Percent = [float]($jobs_done/$total_jobs)
    $Progress = New-BTProgressBar -Status "Backup progress..." -Value $Progress_Percent -ValueDisplay "$jobs_done/$total_jobs"
    New-BurntToastNotification @toastParams_Backup_Progress -Text ("Current backup job:"), ($current_job_name), 
($timestamp).ToString() -ProgressBar $Progress -SuppressPopup 
}

```

So, with your way to reuse the Databinding, how can I insert it before the foreach loop at first, than update these:
```
DataBinding = @{
    FirstLine = "Current backup job:"
    SecondLine = $current_job_name
    ThirdLine = ($timestamp).ToString()
    ProgressStatus = "Backup progress..."
    ProgressValue = $Progress_Percent
    ProgressDisplay = $jobs_done/$total_jobs
}
```
the values are not known on first toast launch !
I also guess DataBinding must be set before the foreach loop than I use something like 
```
if ($first_job) { New-BurntToastNotification ...}
else { Update-BTNotification... }
```

Any hints so that I can easily find a fix ? I will try some samples tomorrow and will give feedback when I am done with a working solution

Best regards

**Reply by Windos** [reply]
I can't dive in too deep atm, but one thing you can use to see if there is already a toast is `Get-BTHistory`:

```powershell
Get-BTHistory -UniqueIdentifier $TargetNoticationfName
```

If you're not sure on progress bar values on first launch, you could potentially do something like:

1. Launch initial toast without progress bar (New-BurntToastNotification)
1. Replace initial toast with a new one with progress bar (New-BurntToastNotification)
1. Update new progress bar toast with updated progress info (Update-BTNotification)

**Comment by PhilZ-cwm6** [comment]
I got it with Get-BTHistory
That way, if the notif was dismissed, i can avoid Update-BTNotification failing because no toast was found

I will look at it in 1-2 days I think and give a feedback
thank you again for the support

**Comment by PhilZ-cwm6** [comment, marked as answer]
Your $Progress tip to use the DataBinding worked perfectly, many thanks !
However, I could not use the get-bthistory as I wanted:

Here's a sample test of what I am trying to achieve and that doesn't work.
When I try to query the contents of the toast, the Contents.InnerText values are not reported properly when using the DataBinding:

```
$DataBinding = @{
    FirstLine = "Current backup job:"
    SecondLine = "line2"
    ThirdLine = "line3"
}

$toastParams_Backup_Progress = @{
    UniqueIdentifier = "My Toast ID"
    Header = (New-BTHeader -Id 1 -Title "My Toast Title")
}

New-BurntToastNotification @toastParams_Backup_Progress -Text "FirstLine", "SecondLine", "ThirdLine" -DataBinding $DataBinding
Start-Sleep -Seconds 3
$toast = get-bthistory -UniqueIdentifier "My Toast ID"
$toast_exists = $toast."Content"."InnerText".Contains("line2")
$toast_exists
```

$toast_exists is always false because $toast."Content"."InnerText" only has this:
` {FirstLine}{SecondLine}{ThirdLine}`

The entries are not populated and I tried every way to query their values, no success

When not using DataBinding, it works perfectly:
```
$toastParams = @{
    UniqueIdentifier = "My Toast ID"
    Header = (New-BTHeader -Id 1 -Title "My Toast Title")
}

$Progress = New-BTProgressBar -Status "Backup progress..." -Value 0.5 -ValueDisplay "3/6"
New-BurntToastNotification @toastParams -Text "line1", "line2", "line3" -ProgressBar $Progress
Start-Sleep -Seconds 3
$toast = get-bthistory -UniqueIdentifier "My Toast ID"
$toast_exists = $toast."Content"."InnerText".Contains("line1")
$toast_exists
```
now, it is true

Also, in both cases, I cannot query the `ProgressStatus` field content

Is there a way at least to fix the first issue: query the text lines contents when using DataBinding ?
Through my script, I use the same UniqueIdentifier for all the job tasks: mount drive, ping network, wait and start for encryption process, error info... So, I need to ensure I have a current progress toast displayed before updating its contents

Many thanks and best regards

**Reply by Windos** [reply]
That object returned by `Get-BTHistory` is a little interesting. The content can be thought of as the "template" for the toast. When using data bindings the template includes "anchors" to bind that data to (the "{FirstLine}" etc.) but when not using a binding the template just has the raw text (usually, things changed between versions and I don't recall offhand what the current behaviour is 😂)

Luckily, that object also includes the data provided by your databinding:

```powershell
$toast.data.values
```

```output
Key             Value              
---             -----              
ProgressValue   0.1                
ProgressDisplay 1/10               
ThirdLine       line2              
ProgressStatus  Backup progress... 
SecondLine      line1              
FirstLine       Current backup job:
```

That's great, but it's a weird object type and I can't figure out how to query specific values directly. Instead, you can copy the data into a hashtable (or some other collection I guess) and query that:

```powershell
$ToastData = @{}

for ($i = 0; $i -lt $toast.data.Values.Key.Count; $i++) {
    $ToastData[$toast.data.Values.Key[$i]] = $toast.data.Values.Value[$i]
}

$ToastData.ProgressStatus
```

```output
Backup progress...
```

**Comment by PhilZ-cwm6** [comment]
I get it now, was driving me crazy since no documentation is present for the Get-BTHistory and I found no examples :-)
I did not think to go into the data.values part of the object and I was focusing on the InnerText and parts with xml

Wired object in fact, seems like a simple table only object
Many thanks for all the hints. I will try to enhance my script in the next couple of days to update the notifications through my code, just for the fun :-)

Best regards

**Reply by Windos** [reply]
And thanks for raising this discussion. It's this sort of thing that helps me know what sort of documentation I should be focusing on (because there's a LOT needed) and better understand how people are using this module (which informs what to focus development time on)

**Comment by Windos** [comment]
Hello, and thank you for your contribution to BurntToast.

Today marks the official release of [BurntToast v1.0.0](https://github.com/Windos/BurntToast/releases/tag/v1.0.0). As part of this major milestone and to ensure the project starts its next chapter with a clean slate, I am declaring "issue bankruptcy."

This decision is explained in more detail in the official announcement post: [Ten Years Toasting: BurntToast Hits the Big v1!](https://toastit.dev/2025/07/22/burnttoast-v1/)

As a result, I am closing all outstanding issues and pull requests (and discussions), including this one. Please do not interpret this as a lack of interest in the problem, suggestion, or topic you've raised. This is a necessary step to clear a very old backlog and focus development efforts on the current state of the module.

**If this topic is still relevant to you:**

1.  Please **open a new discussion**. You are welcome to link back to this one for historical context.

I hope this reset will allow me to be more responsive to community feedback going forward.

Thank you for your understanding and for being part of the BurntToast journey.
